File: tests/fixtures/lua/edge.lua
-- edge.lua — boundary behaviors: metamethod classification in table constructors,
-- anonymous function assignment as Function (not Var), multi-level dot scope,
-- nested table constructor scope, body comments in signature,
-- positional field exclusion, function-body definition exclusion

-- ── Table constructor metamethod classification ──
local mt = {
    __add = function(a, b) return a + b end,
    __index = function(t, k) return rawget(t, k) end,
    __newindex = function(t, k, v) rawset(t, k, v) end,
    __gc = function(t) print("gc") end,
    __tostring = function(t) return tostring(t) end,
    __call = function(t, ...) return t end,
    helper = function() return 1 end,
}


-- ── Anonymous function assignment as Function (not Var) ──
local generate_id = function()
    return math.random(1, 1000000)
end


global_handler = function(event)
    return event
end


-- ── Multi-level dot scope ──
function app.services.queue_task(task)
    return task
end


-- ── Nested table constructor scope ──
local config = {
    db = {
        host = "localhost",
        port = 5432,
    },
    server = {
        name = "api",
    },
}


-- ── Body comments in signature ──
function greet()
    -- This comment should not appear in the signature
    local msg = "hello"
    return msg
end


function calculate(a, b)
    -- Compute the result
    return a + b
end


-- ── Positional field not extracted ──
local items = {
    "positional",
    key = "value",
}


-- ── Function-body definitions NOT extracted ──
function setup()
    local temp = "inside"
    status = "done"
end

File: tests/fixtures/python/edge.py
# edge.py — boundary behaviors: function-body exclusion, dunder at module level,
# Enum/Protocol classification edge cases, TypeAlias annotated form, false-positive resilience,
# Field vs Var vs Alias distinction, same-name disambiguation, multiline signature with comments

# ── Dunder at module level stays Function (not Constructor/Operator) ──
def __init__():
    pass


def __add__(x, y):
    pass


# ── Function-body definitions NOT extracted ──
def factory():
    local_var = 1
    count: int = 0
    def inner():
        pass


# ── TypeAlias annotated form ──
HeaderValue: TypeAlias = str | list[str]


# ── TypeAlias in class NOT Field ──
class AppConfig:
    Value: TypeAlias = dict[str, object]


# ── Enum classification edge cases ──
class Status(SomeMixin, IntEnum):
    OK = 0


class Perm(Flag):
    R = 1


# ── Protocol classification edge cases ──
class MyProto(SomeBase, Protocol):
    def method(self): ...


# ── Enum priority over Protocol ──
class Weird(Enum, Protocol):
    A = 1


# ── Non-Enum/Protocol base → stays Class ──
class ChildClass(Parent):
    def method(self):
        pass


# ── Plain annotation NOT TypeAlias → Var ──
name: str = "hello"


# ── Plain assignment NOT TypeAlias → Var ──
MAX_SIZE = 100


# ── False-positive resilience ──
# tree-sitter AST handles: string/comment/triple-quote never parsed as def/class
def real_func():
    pass


code_string = "def fake_in_string():\n    pass"
# class FakeInComment: pass
doc = '''
def fake_in_triple_quotes():
    pass
class AlsoFake:
    pass
'''


# ── Minimal one-line class ──
class MinimalClass: pass


# ── Multiline signature with inline comments ──
def process_data(
    items: list,  # input items
    verbose: bool,  # verbose mode
) -> None:
    pass


# ── Decorated class with comment ──
@dataclass  # auto-generate __init__
class UserRecord:
    name: str
    age: int


# ── Decorated function with comment on decorator line ──
@retry(max_attempts=3)  # retry up to 3 times
def fetch_url(url: str) -> bytes:
    pass


# ── Stacked decorators with comments ──
@cache  # cache the result
@timeout(30)  # 30 second timeout
def compute(x: int) -> int:
    return x


# ── Multiline type alias ──
type Matrix = list[
    list[float],
]